File: backend/app/api/employee_training.py
"""Employee training API for cohort calendar and materials views."""

from fastapi import APIRouter, Depends
from sqlalchemy.ext.asyncio import AsyncSession

from ..db.database import get_session
from ..db.models import User, UserRole
from ..dependencies import RoleChecker
from ..schemas.employee_training_schema import (
    EmployeeTrainingCalendarResponse,
    EmployeeTrainingMaterialsResponse,
    EmployeeAttendanceHistoryResponse,
)
from ..services.cohort_service import CohortService


employee_training_router = APIRouter()


@employee_training_router.get("/calendar", response_model=EmployeeTrainingCalendarResponse)
async def get_employee_training_calendar(
    current_user: User = Depends(RoleChecker([UserRole.EMPLOYEE])),
    session: AsyncSession = Depends(get_session),
):
    cohorts = await CohortService.list_training_cohorts_for_employee(
        user_id=current_user.id,
        session=session,
    )

    sessions_payload = []
    for cohort in cohorts:
        for item in cohort.sessions:
            sessions_payload.append(
                {
                    "id": item.id,
                    "cohort_id": cohort.id,
                    "cohort_name": cohort.name,
                    "call_title": cohort.call.title if cohort.call else "",
                    "course_title": cohort.course.title if cohort.course else "",
                    "title": item.title,
                    "session_date": item.session_date,
                    "start_time": item.start_time.strftime("%H:%M"),
                    "end_time": item.end_time.strftime("%H:%M"),
                    "location": item.location,
                    "professor_name": (
                        item.professor.user.fullname
                        if item.professor and item.professor.user
                        else ""
                    ),
                }
            )

    sessions_payload.sort(key=lambda s: (s["session_date"], s["start_time"]))
    return {"sessions": sessions_payload, "total": len(sessions_payload)}


@employee_training_router.get("/materials", response_model=EmployeeTrainingMaterialsResponse)
async def get_employee_training_materials(
    current_user: User = Depends(RoleChecker([UserRole.EMPLOYEE])),
    session: AsyncSession = Depends(get_session),
):
    materials = await CohortService.list_training_materials_for_employee(
        user_id=current_user.id,
        session=session,
    )
    return {"materials": materials, "total": len(materials)}


@employee_training_router.get("/attendance", response_model=EmployeeAttendanceHistoryResponse)
async def get_employee_attendance_history(
    current_user: User = Depends(RoleChecker([UserRole.EMPLOYEE])),
    session: AsyncSession = Depends(get_session),
):
    history = await CohortService.list_attendance_history_for_employee(
        user_id=current_user.id,
        session=session,
    )
    return {"attendance": history, "total": len(history)}

File: frontend/lib/employeeTraining.ts
import { apiClient } from "./api";

export interface EmployeeTrainingSession {
  id: number;
  cohort_id: number;
  cohort_name: string;
  call_title: string;
  course_title: string;
  title: string;
  session_date: string;
  start_time: string;
  end_time: string;
  location?: string | null;
  professor_name: string;
}

export interface EmployeeTrainingMaterial {
  id: number;
  cohort_id: number;
  cohort_name: string;
  course_id: number;
  course_title: string;
  title: string;
  description?: string | null;
  file_name: string;
  file_size: number;
  file_type: string;
  created_at: string;
}

export interface EmployeeTrainingCalendarResponse {
  sessions: EmployeeTrainingSession[];
  total: number;
}

export interface EmployeeTrainingMaterialsResponse {
  materials: EmployeeTrainingMaterial[];
  total: number;
}

export type EmployeeAttendanceStatus = "present" | "absent" | "late";

export interface EmployeeAttendanceHistoryItem {
  session_id: number;
  session_title: string;
  session_date?: string;
  start_time: string;
  end_time: string;
  location?: string | null;
  cohort_id: number;
  cohort_name: string;
  course_title: string;
  professor_name: string;
  status: EmployeeAttendanceStatus;
  notes?: string | null;
  marked_at: string;
}

export interface EmployeeAttendanceHistoryResponse {
  attendance: EmployeeAttendanceHistoryItem[];
  total: number;
}

export async function getEmployeeTrainingCalendar(): Promise<EmployeeTrainingCalendarResponse> {
  return apiClient.get<EmployeeTrainingCalendarResponse>("/employee/training/calendar", true);
}

export async function getEmployeeTrainingMaterials(): Promise<EmployeeTrainingMaterialsResponse> {
  return apiClient.get<EmployeeTrainingMaterialsResponse>("/employee/training/materials", true);
}

export async function getEmployeeAttendanceHistory(): Promise<EmployeeAttendanceHistoryResponse> {
  return apiClient.get<EmployeeAttendanceHistoryResponse>("/employee/training/attendance", true);
}
